### ISSUE TITLE
Mention build requirements in README.md

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#465]
README.md doesn't say anything about software requirements for the build machine. Among other things, it doesn't say which version of gcc is supported, and why (it also doesn't mention that we use C++14 or C++17). This information is important to users (see issue #412 where a user was surprised that gcc 5.4 doesn't work), but also to developers - sometimes certain code constructs don't work on certain older compilers so need to be avoided, and we need to know just how old we **want** to support. This knowledge is important to keep developers grounded in reality, not their bleeding-edge compiler installation.

Bout `install-dependencies.sh` and `configure.py` could also confirm that the compiler is new enough. configure.py already checks the installed Boost version, for example.

**Comment by senior7515** [comment]
From personal tests gcc6 fails too. gcc7 or greater works reliably. Tested on Ubuntu/Fedora/Debian

**Comment by nyh** [comment]
I'm not surprised that all of the Seastar developers use bleeding-edge distributions and compiler versions, so it's very easy for code that breaks on gcc 6 to get committed and never fixed. Part of my question is whether this is deliberate - e.g., do we need some specific C++14 feature that doesn't work on gcc 6 - or accidental (there's a tiny, easy-to-work-around, bug in gcc 6 that we tripped on and nobody bothered fixing).

**Comment by senior7515** [comment]
Agreed. Not sure if the CI for this project is avail to the public, but maybe Travis could be the real documentation for this type of behavior. If it passes on Travis is supported. That would of course require all changes to have a travis hook of some sort.

**Comment by abigagli** [comment]
Well, I think it's even worse than just breaking older compiler support. It seems more than this is locked to a specific compiler at the moment (gcc-7). I've just tested with gcc-8.1 and the build fails again with what seems prett much the same error mentioned in #412. Difficult to understand if the code is valid and gcc has a regression, or the gcc-7 is less strict the both gcc-6 and gcc-8...

**Comment by avikivity** [comment]
We don't want to lock to a specific compiler version, but neither do we want to support old toolchains indefinitely. One option is to commit to support the last two (or three) gcc major releases. corresponding to two (or three) years. Currently those are gcc 8, 7 (and maybe 6).

In any case we'll soon want to drop C++14 support.

If there's a regression, please post a fix.

**Comment by denesb** [comment]
I'm on gcc-8.1 and seastar compiles fine for me, though I only compile `libseastar.a`.

**Comment by abigagli** [comment]
Well, I just found out that compilation succeeds if I compile for C++17. If I don't explicitly specify anything (and I assume that makes g++ use gnu++14) then I get errors like in #412.

Nonetheless, I'm now getting linker errors when building apps:
`core/reactor.cc:4027: undefined reference to `YAML::LoadFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
core/reactor.cc:4029: undefined reference to `YAML::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'`

Seems like the libyaml-cpp on my system (Debian Jessie) has been compiled with a too old gcc and I'm hitting ABI compatibility problems...
But yes, as far as non-executables go, by forcing use of C++17, build is fine with gcc-8.1

**Comment by nyh** [comment]
The current documentation of this is much better than it used to be. We now have in README.md:

```
#### Seastar's C++ standard: C++20 or C++23

Seastar supports both C++20, and C++23. The build defaults to the latest
standard supported by your compiler, but can be explicitly selected with
the `--c++-standard` configure option, e.g., `--c++-standard=20`,
or if using CMake directly, by setting on the `CMAKE_CXX_STANDARD` CMake
variable.

See the [compatibity statement](doc/compatibility.md) for more information.
```

And in doc/compatibility.md we have a lot of details, including:
```
Compilers
=========

Seastar supports GCC and Clang. Ports to other compilers are
welcome.

The last two major releases of a compiler are supported (e.g.
GCC 13 and GCC 14). Patches to support older versions are welcome,
as long as they don't require onerous compromises.
```

So I think in some sense this issue is solved. But personally, I would have liked to see at least some of the information on the build requirements - including the compiler version - in README.md itself without needing to find separate documents.

**Comment by nyh** [comment]
By the way, the mention of GCC 13 and 14 is probably out of date, since GCC 15 is already out. On the other hand, two GCC major releases can be released over just a single year - perhaps we should change this rule to supporting 3 releases, not two?

By the way, which gcc release or releases do we check in CI?
